### ISSUE TITLE
Why does JwtDecoder.decode(yourToken) return a map instead of an object?

### ISSUE THREAD

**Issue by LasseRosenow** [gustavo0197/dart-jwt-decoder#10]
JwtDecoder.decode(yourToken) returns a `Map` of `<String, dynamic>` which is not type safe.

I think this is a perfect example for a situation, where an object should be used instead of a map, as the amount of properties that are to be returned are known.

This would make working with this library much easier.

**Comment by 73VW** [comment]
> I think this is a perfect example for a situation, where an object should be used instead of a map, as the amount of properties that are to be returned are known.

I don't agree with this.

According to [jwt documentation (`payload > private claims`)](https://jwt.io/introduction) and [RFC 7519, section 4.3](https://datatracker.ietf.org/doc/html/rfc7519#section-4.3) you are allowed to use any claim you want in your token. 

That means that there is no indication of the number of properties making it impossible to predict and to build an object.

**Comment by luckv** [comment]
Private claims can't collide with registered claims, which are defined by the rfc. Those claims could be parsed in an object, and optionally the raw decoded token can be exposed in the same object.

**Comment by 73VW** [comment]
@luckv totally agree with this for the registered claims.

However by doing this you would still have a `Map` of `<String, dynamic>` for the private claims which are still not type safe.

The best way for me would be to use a custom "Token" class with a factory constructor `fromJson` and pass it as optional parameter to `JwtDecoder.decode(yourToken, fromJson)`. 

Using mechanisms such as [Flutter serialization](https://docs.flutter.dev/development/data-and-backend/json) would make the process of token decoding safer.

**Comment by luckv** [comment]
You are right. Doing a partial serialization can't be useful for all cases, and adding a dependency to a serialization library seems a bit too much for a library that should remain simple.

**Comment by 73VW** [comment]
@luckv I don't think adding a dependency to a serialization library is the way to do this. 

I would suggest adding an option to pass a `Function` to `JwtDecoder.decode` and leave the responsibility of this function to the user.

For me there would be 2 decode methods. 

The one existing now and another with the following signature:

`static T decodeObj<T>(String token, T Function(Map<String, dynamic>) objectDecoder)`

That would allow to use any method/mechanism to obtain an object from a token. I would though strongely recommend using Flutter serialization as it really safe.

What do you think about this?

**Comment by luckv** [comment]
Passing the callback would be the same same as calling it with the decoded jwt already returned.

**Comment by 73VW** [comment]
@luckv any other idea?

**Comment by luckv** [comment]
No, I would leave as it is. It's the best way to decode the token without doing useless computation. 
At most an object with the registered claims and the raw token decoded as map.
